test-analysis-core/src/main/kotlin/com/saveourtool/save/test/analysis/api/TestId.kt

```diff
@@ -0,0 +1,13 @@
+package com.saveourtool.save.test.analysis.api
+
+/**
+ * A test executed within a particular project (the same test executed in a
+ * different project even within the same organization will have a different
+ * [hash]).
+ *
+ * To create a [TestId] instance, use [TestIdGenerator.testId].
+ *
+ * @property hash the SHA hash that uniquely identifies this test.
+ */
+@JvmInline
+value class TestId internal constructor(private val hash: String)
```

Comment:
Maybe call it `TestExecutionId` in order not to confuse with `Test.id`, because it has the same meaning as `TestExecution.id`? Also, why can't we use `TestExecution.id` as a key here?

Reply:
No, this is not a `TestExecutionId`.
This is actually similar to `com.saveourtool.save.entities.Test` (which has a `hash`), but keep reading.

The same `save-cli` test (Git source with a specific version, test suite, TOML file) executed multiple times (in different executions) will have the same `TestId`.

In other words, `TestId` identifies a series of test runs (so that they can be statistically analyzed) rather than a single test. 

Moreover, as you may have noticed, the same test executed for a different project (e.g.: _KtLint_ vs _Diktat_, or _Diktat 1.2.3_ vs _Diktat 1.2.4_) or within a different organization will have a different `TestId`.

For example, on the production database, 

```sql
select count(distinct(test_id)) from test_execution;
```

yields **28**, despite the same [standard test suite](https://github.com/saveourtool/save-cli/tree/main/examples/kotlin-diktat) (mirrored [here](https://github.com/0x6675636b796f75676974687562/save-kotlin-diktat)) has about 22 tests which are executed for **multiple** SAVE-Cloud projects (Diktat, Diktat-Integration, etc.).

So if `Test.hash` were the entity I needed, I would expect a `test_id` count which is equal to or slightly exceeds a **multiple** of 22, e.g. 44+ (at least), 66+, 88+ etc.

In terms of CI and unit testing, my intention was to create something like a JUnit/TestNG _test name_ (e. g.:  [`org.apache.commons.lang3.AnnotationUtilsTest.testEquivalence`](https://gitbox.apache.org/repos/asf?p=commons-lang.git;a=blob_plain;f=src/test/java/org/apache/commons/lang3/AnnotationUtilsTest.java;hb=HEAD)), partitioned by organization name and project name.

Reply:
Awesome explanation, thanks

Reply:
It seems actually possible to use a triplet of `(Test.hash, OrganizationName, ProjectName)`, but this is merely a matter of implementing a yet another `com.saveourtool.save.test.analysis.api.TestIdGenerator`.

Speaking of naming, the entity is actually very similar to `Test.hash` and this may well be retain the `TestId` name. If you don't like that too much, how about renaming it to `TestNameId`?

Reply:
No, I think `TestId` is fine in a sense that it's a unique ID of a test and `TestNameId` would also be confusing with `Test.name`. Maybe `TestExecutionSeriesId` would be more appropriate, but we can keep it simpler.